src/coprocessor/dag/aggr_fn/mod.rs

```diff
@@ -0,0 +1,361 @@
+}
+
+impl<T: Evaluable, State> AggrFunctionStateUpdatePartial<T> for State
+where
+    State: ConcreteAggrFunctionState,
+{
+    // All `ConcreteAggrFunctionState` implement `AggrFunctionStateUpdatePartial<T>`, which is
+    // one of the trait bound that `AggrFunctionState` requires.
+
+    #[inline]
+    default fn update(&mut self, _ctx: &mut EvalContext, _value: &Option<T>) -> Result<()> {
+        panic!("Unmatched parameter type")
+    }
+
+    #[inline]
+    default fn update_repeat(
+        &mut self,
+        _ctx: &mut EvalContext,
+        _value: &Option<T>,
+        _repeat_times: usize,
+    ) -> Result<()> {
+        panic!("Unmatched parameter type")
+    }
+
+    #[inline]
+    default fn update_vector(
+        &mut self,
+        _ctx: &mut EvalContext,
+        _values: &[Option<T>],
+    ) -> Result<()> {
+        panic!("Unmatched parameter type")
+    }
+}
+
+impl<T: Evaluable, State> AggrFunctionStateUpdatePartial<T> for State
+where
+    State: ConcreteAggrFunctionState<ParameterType = T>,
+{
+    #[inline]
+    fn update(&mut self, ctx: &mut EvalContext, value: &Option<T>) -> Result<()> {
+        self.update_concrete(ctx, value)
+    }
+
+    #[inline]
+    fn update_repeat(
+        &mut self,
+        ctx: &mut EvalContext,
+        value: &Option<T>,
+        repeat_times: usize,
+    ) -> Result<()> {
+        for _ in 0..repeat_times {
+            self.update_concrete(ctx, value)?;
+        }
+        Ok(())
+    }
+
+    #[inline]
+    fn update_vector(&mut self, ctx: &mut EvalContext, values: &[Option<T>]) -> Result<()> {
+        for value in values {
+            self.update_concrete(ctx, value)?;
+        }
+        Ok(())
+    }
+}
+
+impl<T: AggrResultAppendable + ?Sized, State> AggrFunctionStateResultPartial<T> for State
+where
+    State: ConcreteAggrFunctionState,
+{
+    // All `ConcreteAggrFunctionState` implement `AggrFunctionStateResultPartial<T>`, which is
+    // one of the trait bound that `AggrFunctionState` requires.
+
+    #[inline]
+    default fn push_result(&self, _ctx: &mut EvalContext, _target: &mut T) -> Result<()> {
+        panic!("Unmatched result append target type")
+    }
+}
+
+impl<T: AggrResultAppendable + ?Sized, State> AggrFunctionStateResultPartial<T> for State
+where
+    State: ConcreteAggrFunctionState<ResultTargetType = T>,
+{
+    #[inline]
+    fn push_result(&self, ctx: &mut EvalContext, target: &mut T) -> Result<()> {
+        self.push_result_concrete(ctx, target)
+    }
+}
+
+impl<F> AggrFunctionState for F
+where
+    F: ConcreteAggrFunctionState,
+{
+    // All `ConcreteAggrFunctionState` can implement `AggrFunctionState` now, since they meet
+    // all trait bound of `AggrFunctionState`.
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_type_match() {
+        /// A state that accepts Int and outputs Real.
+        #[derive(Clone, Debug)]
+        struct AggrFnStateFoo {
```

Comment:
Could we implement an aggregation function directly in this PR as a sample so we needn't implement one in the test?

Reply:
I intentionally extracted them into different PRs because a fully functional aggr function implementation is not very short 🙃 For a real implementation, see  https://github.com/tikv/tikv/blob/ab507abeee4c809ea30d8bfbcc37304416a0c537/src/coprocessor/dag/aggr_fn/impl_avg.rs#L99 (notice that the parser part is not included in this PR)